### PR TITLE
[Oracle][MySQL][SqlServer][PostgresSQL] Fix Oracle/MySQL/SqlServer/PostgresSQL CDC parser schema change event failed

### PR DESCRIPTION
**Pull request by e-mhui**
Fix: #2055 #1792 #1379 #1807 #1810 #1803 #1804 #2311 

When a SchemaChangeEvent is triggered, the `schemaChangeRecordValue` will obtain binlog information from **sourceInfo**. However, only MySQL CDC has binlog information, and other CDC will report an error after triggering a SchemaChangeEvent event.

image

So I provided an interface for each connector to implement their own sourceinfo parsing.

```java
/** SchemaChangeEvent Handler. */
public interface SchemaChangeEventHandler {
    Map<String, Object> parseSource(SchemaChangeEvent event);
}
```

Then, it's used in JdbcSourceEventDispatcher#schemaChangeRecordValue.

```java
        private Struct schemaChangeRecordValue(SchemaChangeEvent event) throws IOException {
            Map<String, Object> source = schemaChangeEventHandler.parseSource(event);
            HistoryRecord historyRecord =
                    new HistoryRecord(
                            source,
                            event.getOffset(),
                            event.getDatabase(),
                            null,
                            event.getSchema(),
                            event.getDdl(),
                            event.getTableChanges());
        // ... other code ... 
```

For MySQL CDC

```java
/** MySqlSchemaChangeEventHandler to deal MySql's schema change event. */
public class MySqlSchemaChangeEventHandler implements SchemaChangeEventHandler {

    @Override
    public Map<String, Object> parseSource(SchemaChangeEvent event) {
        Map<String, Object> source = new HashMap<>();
        Struct sourceInfo = event.getSource();
        String fileName = sourceInfo.getString(BINLOG_FILENAME_OFFSET_KEY);
        Long pos = sourceInfo.getInt64(BINLOG_POSITION_OFFSET_KEY);
        Long serverId = sourceInfo.getInt64(SERVER_ID_KEY);
        source.put(SERVER_ID_KEY, serverId);
        source.put(BINLOG_FILENAME_OFFSET_KEY, fileName);
        source.put(BINLOG_POSITION_OFFSET_KEY, pos);
        return source;
    }
}
```

For Oracle CDC

```java
/** OracleSchemaChangeEventHandler to deal Oracle's schema change event. */
public class OracleSchemaChangeEventHandler implements SchemaChangeEventHandler {

    @Override
    public Map<String, Object> parseSource(SchemaChangeEvent event) {
        Map<String, Object> source = new HashMap<>();
        Struct sourceInfo = event.getSource();
        String scn = sourceInfo.getString(SCN_KEY);
        source.put(SCN_KEY, scn);
        return source;
    }
}
```

### CODE REVIEW

**Comment by e-mhui** [comment]
@ruanhang1993 pls review it.

**Review by ruanhang1993 — COMMENTED** [review]
@e-mhui Thanks for the work.
Maybe Oracle cdc should provide a new dispatcher by self instead of modifying the dispatcher in connector-base. WDYT?
Maybe we could change some scope of methods for JdbcSourceEventDispatcher to reuse its code.

File: flink-cdc-base/src/main/java/com/ververica/cdc/connectors/base/utils/SourceRecordUtils.java [hunk 1]
```diff
@@ -44,10 +44,16 @@ public class SourceRecordUtils {
 
     private SourceRecordUtils() {}
 
-    public static final String SCHEMA_CHANGE_EVENT_KEY_NAME =
+    public static final String MYSQL_SCHEMA_CHANGE_EVENT_KEY_NAME =
```
Comment (ruanhang1993) [inline flink-cdc-base/src/main/java/com/ververica/cdc/connectors/base/utils/SourceRecordUtils.java:47]: Could you help to refactor this code ? The information about mysql should not be found in base module.

Comment (e-mhui) [reply flink-cdc-base/src/main/java/com/ververica/cdc/connectors/base/utils/SourceRecordUtils.java:47]: I try it.

**Comment by e-mhui** [comment]
> @e-mhui Thanks for the work.
> Maybe Oracle cdc should provide a new dispatcher by self instead of modifying the dispatcher in connector-base. WDYT?
> Maybe we could change some scope of methods for JdbcSourceEventDispatcher to reuse its code.

@ruanhang1993 Hi, I have refactored the code, please review it again.

1. Use the `schemaChangeEventHandler.parseSource(event)` interface to handle different CDC sources. The construction of the source mainly refers to the following links:

- https://debezium.io/documentation/reference/1.9/connectors/mysql.html#mysql-schema-change-topic
- https://debezium.io/documentation/reference/stable/connectors/oracle.html#oracle-schema-change-topic
- https://debezium.io/documentation/reference/1.9/connectors/sqlserver.html#about-the-debezium-sqlserver-connector-schema-change-topic
- There is no schema-change-topic for PostgreSQL

2. Use `io.debezium.connector.*.SchemaChangeKey` to determine the SchemaChangeEvent.

**Comment by gtk96** [comment]
@e-mhui please run mvn spotless:apply fix code style,thanks

**Comment by gtk96** [comment]
@e-mhui  please refer https://github.com/ververica/flink-cdc-connectors/blob/master/README.md#code-style fix your code style to pass ci check

**Comment by e-mhui** [comment]
> @e-mhui please refer https://github.com/ververica/flink-cdc-connectors/blob/master/README.md#code-style fix your code style to pass ci check

@gtk96 I had been fixed the check style, can you rerun the workflow ?

**Comment by gtk96** [comment]
cc @ruanhang1993

**Comment by e-mhui** [comment]
@ruanhang1993 pls review it.

**Comment by cainbit** [comment]
@ruanhang1993 Could you please help to review this pr? Many thanks.

File: flink-connector-postgres-cdc/src/main/java/com/ververica/cdc/connectors/postgres/source/handler/PostgresSchemaChangeEventHandler.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+import java.util.HashMap;
+import java.util.Map;
+
+/** PostgresSchemaChangeEventHandler to deal Postgres's schema change event. */
+public class PostgresSchemaChangeEventHandler implements SchemaChangeEventHandler {
+
+    @Override
+    public Map<String, Object> parseSource(SchemaChangeEvent event) {
+        return new HashMap<>();
```
[the remark this answers is not in the document]
Comment (e-mhui) [reply flink-connector-postgres-cdc/src/main/java/com/ververica/cdc/connectors/postgres/source/handler/PostgresSchemaChangeEventHandler.java:30]: From the debezium documnet, PG not exist `schema-change-topic` , so we don't parse it.
Other source has `schema-change-topic`.
https://debezium.io/documentation/reference/1.9/connectors/mysql.html#mysql-schema-change-topic
https://debezium.io/documentation/reference/stable/connectors/oracle.html#oracle-schema-change-topic
https://debezium.io/documentation/reference/1.9/connectors/sqlserver.html#about-the-debezium-sqlserver-connector-schema-change-topic

**Comment by sin70611** [comment]
@ruanhang1993  Could you please help to review this pr? Many thanks.

**Comment by donsmile** [comment]
SchemaChangeEventHandler，这个接口在哪个版本里，我找了一下，最新版的flink-cdc-base 3.0.0 没有看到相关代码。是需要自己修改源码自己编译吗

**Comment by klion26** [comment]
Hi @e-mhui do you still working on this issue?  we encounter the following error when using oraclecdc, looking forward to fix this problem

```
2024-01-25 15:36:20,876 ERROR io.debezium.pipeline.ErrorHandler                            [] - Producer failure
com.ververica.cdc.connectors.shaded.org.apache.kafka.connect.errors.DataException: file is not a valid field name
	at com.ververica.cdc.connectors.shaded.org.apache.kafka.connect.data.Struct.lookupField(Struct.java:254) ~[flink-sql-connector-mysql-cdc-3.0.0.jar:3.0.0]
	at com.ververica.cdc.connectors.shaded.org.apache.kafka.connect.data.Struct.getCheckType(Struct.java:261) ~[flink-sql-connector-mysql-cdc-3.0.0.jar:3.0.0]
	at com.ververica.cdc.connectors.shaded.org.apache.kafka.connect.data.Struct.getString(Struct.java:158) ~[flink-sql-connector-mysql-cdc-3.0.0.jar:3.0.0]
	at com.ververica.cdc.connectors.base.relational.JdbcSourceEventDispatcher$SchemaChangeEventReceiver.schemaChangeRecordValue(JdbcSourceEventDispatcher.java:193) ~[flink-sql-connector-oracle-cdc-3.0.0.jar:3.0.0]
	at com.ververica.cdc.connectors.base.relational.JdbcSourceEventDispatcher$SchemaChangeEventReceiver.schemaChangeEvent(JdbcSourceEventDispatcher.java:223) ~[flink-sql-connector-oracle-cdc-3.0.0.jar:3.0.0]
	at io.debezium.connector.oracle.OracleSchemaChangeEventEmitter.emitSchemaChangeEvent(OracleSchemaChangeEventEmitter.java:122) ~[flink-sql-connector-oracle-cdc-3.0.0.jar:3.0.0]
```

**Comment by e-mhui** [comment]
> Hi @e-mhui do you still working on this issue? we encounter the following error when using oraclecdc, looking forward to fix this problem
> 
> ```
> 2024-01-25 15:36:20,876 ERROR io.debezium.pipeline.ErrorHandler                            [] - Producer failure
> com.ververica.cdc.connectors.shaded.org.apache.kafka.connect.errors.DataException: file is not a valid field name
> 	at com.ververica.cdc.connectors.shaded.org.apache.kafka.connect.data.Struct.lookupField(Struct.java:254) ~[flink-sql-connector-mysql-cdc-3.0.0.jar:3.0.0]
> 	at com.ververica.cdc.connectors.shaded.org.apache.kafka.connect.data.Struct.getCheckType(Struct.java:261) ~[flink-sql-connector-mysql-cdc-3.0.0.jar:3.0.0]
> 	at com.ververica.cdc.connectors.shaded.org.apache.kafka.connect.data.Struct.getString(Struct.java:158) ~[flink-sql-connector-mysql-cdc-3.0.0.jar:3.0.0]
> 	at com.ververica.cdc.connectors.base.relational.JdbcSourceEventDispatcher$SchemaChangeEventReceiver.schemaChangeRecordValue(JdbcSourceEventDispatcher.java:193) ~[flink-sql-connector-oracle-cdc-3.0.0.jar:3.0.0]
> 	at com.ververica.cdc.connectors.base.relational.JdbcSourceEventDispatcher$SchemaChangeEventReceiver.schemaChangeEvent(JdbcSourceEventDispatcher.java:223) ~[flink-sql-connector-oracle-cdc-3.0.0.jar:3.0.0]
> 	at io.debezium.connector.oracle.OracleSchemaChangeEventEmitter.emitSchemaChangeEvent(OracleSchemaChangeEventEmitter.java:122) ~[flink-sql-connector-oracle-cdc-3.0.0.jar:3.0.0]
> ```

Yes, it will be fixed in this pr.

File: flink-cdc-base/src/main/java/com/ververica/cdc/connectors/base/relational/handler/SchemaChangeEventHandler.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+ * limitations under the License.
+ */
+
+package com.ververica.cdc.connectors.base.relational.handler;
+
+import io.debezium.schema.SchemaChangeEvent;
+
+import java.util.Map;
+
+/** SchemaChangeEvent Handler. */
```
Comment (ruanhang1993) [inline flink-cdc-base/src/main/java/com/ververica/cdc/connectors/base/relational/handler/SchemaChangeEventHandler.java:23]: Please add `@Experimental` to this interface. Maybe we will change this interface in other versions.

Comment (ruanhang1993) [inline flink-cdc-base/src/main/java/com/ververica/cdc/connectors/base/relational/handler/SchemaChangeEventHandler.java:23]: /** This handler helps to parse the source struct in SchemaChangeEvent and generate source info. */

**Review by ruanhang1993 — COMMENTED** [review]
@e-mhui .Sorry for the late reply. Left some comments.

File: flink-cdc-base/src/main/java/com/ververica/cdc/connectors/base/relational/handler/SchemaChangeEventHandler.java (hunk 3 above)
Comment (e-mhui) [reply flink-cdc-base/src/main/java/com/ververica/cdc/connectors/base/relational/handler/SchemaChangeEventHandler.java]: Thanks, it has done.

Comment (e-mhui) [reply flink-cdc-base/src/main/java/com/ververica/cdc/connectors/base/relational/handler/SchemaChangeEventHandler.java]: Thanks, it has done.l

**Review by ruanhang1993 — APPROVED** [review]
LGTM.

**Comment by Jetol** [comment]
> SchemaChangeEventHandler，这个接口在哪个版本里，我找了一下，最新版的flink-cdc-base 3.0.0 没有看到相关代码。是需要自己修改源码自己编译吗

oracle 表结构变动报错，用到oracle-cdc 3.0包，我这边也报错，请问你这边解决了吗

**Comment by e-mhui** [comment]
> > SchemaChangeEventHandler，这个接口在哪个版本里，我找了一下，最新版的flink-cdc-base 3.0.0 没有看到相关代码。是需要自己修改源码自己编译吗
> 
> oracle 表结构变动报错，用到oracle-cdc 3.0包，我这边也报错，请问你这边解决了吗

This PR will fix it.

**Comment by ruanhang1993** [comment]
@e-mhui Hi. We need to add some tests to test this bug. 
Please tell me when you need to review again. Thanks.

**Comment by e-mhui** [comment]
> @e-mhui Hi. We need to add some tests to test this bug. Please tell me when you need to review again. Thanks.

@ruanhang1993 Hi, pls review again.

**Comment by klion26** [comment]
thanks @e-mhui for contributing and @ruanhang1993 for merging this!

**Comment by stardustman** [comment]
> > > SchemaChangeEventHandler，这个接口在哪个版本里，我找了一下，最新版的flink-cdc-base 3.0.0 没有看到相关代码。是需要自己修改源码自己编译吗
> > 
> > 
> > oracle 表结构变动报错，用到oracle-cdc 3.0包，我这边也报错，请问你这边解决了吗
> 
> This PR will fix it.
I also didn't find the SchemaChangeEventHandler in the flink-connector-base:1.18.1, so how to find and use it?

**Comment by stardustman** [comment]
> > > SchemaChangeEventHandler，这个接口在哪个版本里，我找了一下，最新版的flink-cdc-base 3.0.0 没有看到相关代码。是需要自己修改源码自己编译吗
> > 
> > 
> > oracle 表结构变动报错，用到oracle-cdc 3.0包，我这边也报错，请问你这边解决了吗
> 
> This PR will fix it.

I did't find handler packages in the [/com/ververica/cdc/connectors/base/relational](https://github.com/apache/flink-cdc/tree/release-3.0/flink-cdc-connect/flink-cdc-source-connectors/flink-cdc-base/src/main/java/com/ververica/cdc/connectors/base/relational) and the flink-connector-oracle-cdc-3.0.1 also not found, how to deal with the schema changes myself using the release-3.0?  thanks a lot.@e-mhui

**Comment by e-mhui** [comment]
The latest release 3.0.1 was packaged on Jan 22 and has not fixed this issue. If you want to use it in advance, you can build it yourself from the master branch. Otherwise, you will have to wait for the next release version.

package command:
```
git clone https://github.com/ververica/flink-cdc-connectors.git
cd flink-cdc-connectors
mvn clean install -DskipTests
```

**Comment by stardustman** [comment]
> @stardustman @Jetol The latest release 3.0.1 was packaged on Jan 22 and has not fixed this issue. If you want to use it in advance, you can build it yourself from the master branch. Otherwise, you will have to wait for the next release version.
> 
> package command:
> 
> ```
> git clone https://github.com/ververica/flink-cdc-connectors.git
> cd flink-cdc-connectors
> mvn clean install -DskipTests
> ```

hello, I'm a new user of the flink-cdc, I have get latest package using master branch. and import the dependecyies. will the `OracleSchemaChangeEventHandler` work automatically or I need other configs to enable it? thanks a lot. @e-mhui

**Comment by e-mhui** [comment]
> > @stardustman @Jetol The latest release 3.0.1 was packaged on Jan 22 and has not fixed this issue. If you want to use it in advance, you can build it yourself from the master branch. Otherwise, you will have to wait for the next release version.
> > package command:
> > ```
> > git clone https://github.com/ververica/flink-cdc-connectors.git
> > cd flink-cdc-connectors
> > mvn clean install -DskipTests
> > ```
> 
> hello, I'm a new user of the `flink-cdc`, I have get latest package using master branch. and import the dependecyies. will the `OracleSchemaChangeEventHandler` work automatically or I need other configs to enable it? thanks a lot. @e-mhui

@stardustmanYes, it will work automatically. However, you may need to configure some parameters to capture DDL events.

```sql
debezium.database.history.store.only.captured.tables.ddl = true
```

You can also refer to:
1. https://nightlies.apache.org/flink/flink-cdc-docs-release-3.0/zh/docs/connectors/cdc-connectors/oracle-cdc/ 
2. https://debezium.io/documentation/reference/1.9/connectors/oracle.html#oracle-property-database-history-store-only-captured-tables-ddl

**Comment by stardustman** [comment]
> ```sql
> debezium.database.history.store.only.captured.tables.ddl
> ```

I read the doc [docs/connectors/cdc-connectors/oracle-cdc/#connector-options](https://nightlies.apache.org/flink/flink-cdc-docs-release-3.0/zh/docs/connectors/cdc-connectors/oracle-cdc/#connector-options). is it the config belongs to `debezium` should be started with "debezium" when using the flink cdc? but I find the demo code [cdc-connectors/oracle-cdc/#datastream-source](https://nightlies.apache.org/flink/flink-cdc-docs-release-3.0/zh/docs/connectors/cdc-connectors/oracle-cdc/#datastream-source)
this line `debeziumProperties.setProperty("log.mining.strategy", "online_catalog");` of  the main method, `log.mining.strategy` didn't start with the `debezium`, after read the [log.mining.strategy](https://debezium.io/documentation/reference/1.9/connectors/oracle.html#oracle-property-log-mining-strategy) it seems the property is belongs to debezium, and the `log.mining.strategy` also need to be set `redo_log_catalog` to capture the DDL?

```java
 public static void main(String[] args) throws Exception {
        Properties debeziumProperties = new Properties();
        debeziumProperties.setProperty("log.mining.strategy", "online_catalog"); 

```
will the "log.mining.strategy" need to start with "debezium"? thanks a lot for bothering you many times😄.   @e-mhui

**Comment by e-mhui** [comment]
> > ```sql
> > debezium.database.history.store.only.captured.tables.ddl
> > ```
> 
> I read the doc [docs/connectors/cdc-connectors/oracle-cdc/#connector-options](https://nightlies.apache.org/flink/flink-cdc-docs-release-3.0/zh/docs/connectors/cdc-connectors/oracle-cdc/#connector-options). is it the config belongs to `debezium` should be started with "debezium" when using the flink cdc? but I find the demo code [cdc-connectors/oracle-cdc/#datastream-source](https://nightlies.apache.org/flink/flink-cdc-docs-release-3.0/zh/docs/connectors/cdc-connectors/oracle-cdc/#datastream-source) this line `debeziumProperties.setProperty("log.mining.strategy", "online_catalog");` of the main method, `log.mining.strategy` didn't start with the `debezium`, after read the [log.mining.strategy](https://debezium.io/documentation/reference/1.9/connectors/oracle.html#oracle-property-log-mining-strategy) it seems the property is belongs to debezium, and the `log.mining.strategy` also need to be set `redo_log_catalog` to capture the DDL?
> 
> ```java
>  public static void main(String[] args) throws Exception {
>         Properties debeziumProperties = new Properties();
>         debeziumProperties.setProperty("log.mining.strategy", "online_catalog"); 
> ```
> 
> will the "log.mining.strategy" need to start with "debezium"? thanks a lot for bothering you many times😄. @e-mhui

It is a parameter of debezium, so you should add the prefix `debezium.`.

image

**Comment by stardustman** [comment]
> > SchemaChangeEventHandler，这个接口在哪个版本里，我找了一下，最新版的flink-cdc-base 3.0.0 没有看到相关代码。是需要自己修改源码自己编译吗
> 
> oracle 表结构变动报错，用到oracle-cdc 3.0包，我这边也报错，请问你这边解决了吗

你好，你是实现 DebeziumDeserializationSchema 接口自定义 Deserialization 来接收数据的么? 在这里是怎样区分 DDL  和实际表的数据的呢？有没有脱敏的 demo 可以发一下？谢谢！@Jetol @e-mhui

**Comment by mdzz9527** [comment]
合并到哪个版本了呢

**Comment by shangeyao** [comment]
> 合并到哪个版本了呢

3.1

**Pull request merged**